### PR TITLE
The `send()` method now only throws when the socket is in the `CONNECTING` state

### LINKED ISSUE #377: send should throw an error if readyState is CONNECTING
`.send()` should [throw an error](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/send#exceptions) if `readyState` is `CONNECTING`, but this implementation does not. The function [only throws an error](https://github.com/thoov/mock-socket/blob/master/src/websocket.js#L142-L144) if the `readyState` is `CLOSING` or `CLOSED`.

Thanks!

### PR DESCRIPTION
**Pull request by steveluscher**
Fixes #377 and #381.

### CODE REVIEW

**Comment by steveluscher** [comment]
This is probably a breaking change for `mock-socket`, since other folks are undoubtedly relying on the existing throw behaviour.

File: tests/unit/websocket.test.js [hunk 1]
```diff
@@ -45,14 +45,44 @@ test('that passing protocols into the constructor works', t => {
...
-    'WebSocket is already in CLOSING or CLOSED state',
-    'an expection is thrown when sending while closed'
+    "Failed to execute 'send' on 'WebSocket': Still in CONNECTING state",
+    'an exception is thrown when sending while in the `CONNECTING` state'
+  );
+});
+
+test('that sending when the socket is in the `CLOSING` state does not throw an exception', t => {
+  const mySocket = new WebSocket('ws://not-real', 'foo');
+  mySocket.readyState = WebSocket.CLOSED;
```
Comment (Atrue) [inline tests/unit/websocket.test.js:62]: it's better to simulate it with `close` method with checking the readyState is CLOSING

File: tests/unit/websocket.test.js [hunk 2]
```diff
@@ -45,14 +45,44 @@ test('that passing protocols into the constructor works', t => {
...
+  t.notThrows(
+    () => {
+      mySocket.send('testing');
+    },
+  );
+});
+
+test('that sending when the socket is in the `CLOSED` state does not throw an exception', t => {
+  const mySocket = new WebSocket('ws://not-real', 'foo');
+  mySocket.readyState = WebSocket.CLOSED;
```
Comment (Atrue) [inline tests/unit/websocket.test.js:72]: it's better to simulate it with `close` method and `onclose` subscription with checking the readyState is CLOSED

File: tests/unit/websocket.test.js [hunk 3]
```diff
@@ -45,14 +45,44 @@ test('that passing protocols into the constructor works', t => {
...
+  t.notThrows(
+    () => {
+      mySocket.send('testing');
+    },
+  );
+});
+
+test('that sending when the socket is in the `OPEN` state does not throw an exception', t => {
```
Comment (Atrue) [inline tests/unit/websocket.test.js:80]: not really necessary as this is the base case testing in the most of the test cases

**Review by Atrue — CHANGES_REQUESTED** [review]

File: tests/unit/websocket.test.js (hunk 1 above)
Comment (steveluscher) [reply tests/unit/websocket.test.js:62]: In any case, this is a straight up typo on my part.

File: tests/unit/websocket.test.js [hunk 4]
```diff
@@ -45,14 +45,37 @@ test('that passing protocols into the constructor works', t => {
...
-    'an expection is thrown when sending while closed'
+    "Failed to execute 'send' on 'WebSocket': Still in CONNECTING state",
+    'an exception is thrown when sending while in the `CONNECTING` state'
   );
 });
+
+test('that sending when the socket is in the `CLOSING` state does not throw an exception', t => {
+  const mySocket = new WebSocket('ws://not-real', 'foo');
+  mySocket.close();
+  t.notThrows(
```
Comment (Atrue) [inline tests/unit/websocket.test.js:63]: please add `t.is(mySocket.readyState, WebSocket.CLOSING)` to make sure the case testing the expected behaviour

File: tests/unit/websocket.test.js [hunk 5]
```diff
@@ -45,14 +45,37 @@ test('that passing protocols into the constructor works', t => {
...
+      mySocket.send('testing');
+    },
+  );
+});
+
+test.cb('that sending when the socket is in the `CLOSED` state does not throw an exception', t => {
+  const mySocket = new WebSocket('ws://not-real', 'foo');
+  mySocket.close();
+  mySocket.addEventListener('close', () => {
+    t.notThrows(
```
Comment (Atrue) [inline tests/unit/websocket.test.js:74]: please add `t.is(mySocket.readyState, WebSocket.CLOSED)` to make sure the case testing the expected behaviour

File: src/websocket.js [hunk 6]
```diff
@@ -139,8 +139,8 @@ class WebSocket extends EventTarget {
   }
 
   send(data) {
-    if (this.readyState === WebSocket.CLOSING || this.readyState === WebSocket.CLOSED) {
-      throw new Error('WebSocket is already in CLOSING or CLOSED state');
+    if (this.readyState === WebSocket.CONNECTING) {
+      throw new DOMException("Failed to execute 'send' on 'WebSocket': Still in CONNECTING state");
```
Comment (Atrue) [inline src/websocket.js:143]: use the base Error as DOMException does not exist in node

**Review by Atrue — CHANGES_REQUESTED** [review]

File: src/websocket.js (hunk 6 above)
Comment (Atrue) [reply src/websocket.js:143]: please change it to `Error` (the 1st variant) to support current node versions.
But you can also add a comment like `// TODO: node>=17 replace with DOMException`

File: tests/unit/websocket.test.js [hunk 7]
```diff
@@ -45,14 +45,39 @@ test('that passing protocols into the constructor works', t => {
...
+
+test.cb('that sending when the socket is in the `CLOSED` state does not throw an exception', t => {
+  const mySocket = new WebSocket('ws://not-real', 'foo');
+  mySocket.close();
+  mySocket.addEventListener('close', () => {
+    t.is(mySocket.readyState, WebSocket.CLOSED);
+    t.notThrows(
+      () => {
+        mySocket.send('testing');
+        t.end();
```
Comment (Atrue) [inline tests/unit/websocket.test.js:79]: put `t.end` after `t.notThrows` as this is not a part of testing callback

**Review by Atrue — CHANGES_REQUESTED** [review]

File: tests/unit/websocket.test.js (hunk 7 above)
Comment (steveluscher) [reply tests/unit/websocket.test.js:79]: giphy-4

**Review by Atrue — APPROVED** [review]

**Pull request merged**